File: jflove-app/lib/services/note_service.dart
import '../models/note.dart';
import '../utils/http_service.dart';

/// 笔记管理服务
/// 对标 jflove-desktop/src/services/note_service.py
///
/// 服务端接口对照：
///   GET    /api/v1/notes/list   - 笔记列表
///   GET    /api/v1/notes/read   - 读取笔记内容（body: token, filename）
///   POST   /api/v1/notes/write  - 新建或覆盖笔记（body: token, filename, content）
///   DELETE /api/v1/notes/       - 删除笔记（body: token, filename）
///   POST   /api/v1/notes/rename - 重命名（body: token, old_name, new_name）
///   GET    /api/v1/notes/disk-config   - 获取笔记目录配置
///   PUT    /api/v1/notes/disk-config   - 设置笔记目录配置
class NoteService {
  final HttpService _http;

  NoteService(this._http);

  /// 列出所有笔记
  Future<List<Note>> listNotes() async {
    final resp = await _http.encryptedGet('/api/v1/notes/list', {});
    final items = resp['notes'] as List<dynamic>;
    return items.map((e) => Note.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// 读取笔记内容（服务端路径为 /read，不是 /get）
  Future<Note> getNote(String filename) async {
    final resp = await _http.encryptedGet('/api/v1/notes/read', {
      'filename': filename,
    });
    return Note.fromJson(resp);
  }

  /// 新建或覆盖笔记（服务端用 /write 同时处理新建和保存）
  Future<void> createNote(String filename) async {
    await _http.encryptedPost('/api/v1/notes/write', {
      'filename': filename,
      'content': '',
    });
  }

  /// 保存笔记内容（服务端用 POST /write，不是 PUT /save）
  Future<void> saveNote(String filename, String content) async {
    await _http.encryptedPost('/api/v1/notes/write', {
      'filename': filename,
      'content': content,
    });
  }

  /// 重命名笔记（服务端用 POST）
  Future<void> renameNote(String oldName, String newName) async {
    await _http.encryptedPost('/api/v1/notes/rename', {
      'old_name': oldName,
      'new_name': newName,
    });
  }

  /// 删除笔记
  Future<void> deleteNote(String name) async {
    await _http.encryptedDelete('/api/v1/notes', {'filename': name});
  }

  // ── 笔记目录配置 ─────────────────────────────────

  /// 获取当前用户的笔记磁盘配置及可用磁盘列表
  ///
  /// 对应服务端：GET /api/v1/notes/disk-config
  /// 返回：{ disk_id, path, disks }（disks 为可用磁盘列表）
  Future<Map<String, dynamic>> getNotesDiskConfig() async {
    return _http.encryptedGet('/api/v1/notes/disk-config', {});
  }

  /// 设置当前用户的笔记存储磁盘和子目录
  ///
  /// 对应服务端：PUT /api/v1/notes/disk-config
  /// [diskId] 目标磁盘 ID，null 表示清除配置
  /// [path] 磁盘内子目录相对路径，空字符串表示磁盘根目录
  Future<void> setNotesDiskConfig(int? diskId, {String path = ''}) async {
    await _http.encryptedPut('/api/v1/notes/disk-config', {
      'disk_id': diskId,
      'path': path,
    });
  }
}

File: jflove-desktop/ui/src/services/note-service.ts
/**
 * 笔记管理服务
 *
 * 对标桌面端 src/services/note_service.py。
 * 接口路径/字段与后端 note_controller.py 保持一致：
 *   - GET/POST  /api/v1/notes/list       笔记列表
 *   - GET/POST  /api/v1/notes/read       读取内容（body: filename）
 *   - POST      /api/v1/notes/write      新建或覆盖（body: filename, content）
 *   - DELETE    /api/v1/notes/           删除（body: filename）
 *   - POST      /api/v1/notes/rename     重命名（body: old_name, new_name）
 *   - GET/POST  /api/v1/notes/disk-config 获取笔记磁盘配置
 *   - PUT       /api/v1/notes/disk-config 设置笔记磁盘配置
 * 注：http-client.get 内部用 POST 发送（浏览器 GET 不能携带 body）。
 */

import { httpClient } from '../utils/http-client';
import type { Note } from '../types/models';

export const noteService = {
  /** 获取笔记列表 */
  async listNotes(): Promise<Note[]> {
    const resp = await httpClient.get<{ notes: Note[] }>('/api/v1/notes/list');
    return resp.notes;
  },

  /** 获取笔记内容 */
  async getNote(filename: string): Promise<string> {
    const resp = await httpClient.get<{ content: string }>(
      '/api/v1/notes/read',
      { filename },
    );
    return resp.content;
  },

  /** 保存笔记（覆盖已有内容） */
  async saveNote(filename: string, content: string): Promise<void> {
    await httpClient.post('/api/v1/notes/write', {
      filename,
      content,
    });
  },

  /** 创建笔记（后端通过 write 接口新建，空内容） */
  async createNote(filename: string): Promise<void> {
    await httpClient.post('/api/v1/notes/write', {
      filename,
      content: '',
    });
  },

  /** 重命名笔记 */
  async renameNote(oldFilename: string, newFilename: string): Promise<void> {
    await httpClient.post('/api/v1/notes/rename', {
      old_name: oldFilename,
      new_name: newFilename,
    });
  },

  /** 删除笔记 */
  async deleteNote(filename: string): Promise<void> {
    await httpClient.delete('/api/v1/notes/', {
      filename,
    });
  },

  /** 获取当前用户的笔记磁盘配置 */
  async getNotesDiskConfig(): Promise<{
    disk_id: number | null;
    path: string;
    disks: { id: number; name: string }[];
  }> {
    return httpClient.get('/api/v1/notes/disk-config');
  },

  /** 设置当前用户的笔记磁盘配置 */
  async setNotesDiskConfig(diskId: number | null, path = ''): Promise<void> {
    await httpClient.put('/api/v1/notes/disk-config', {
      disk_id: diskId,
      path,
    });
  },
};
